#! /bin/sh -e

# All lines beginning with `# DPATCH:' are a description of the patch.
# DP: correctly limit reload class
# DP: fixes 375522

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p1 < $0
        #cd ${dir}gcc && autoconf
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p1 < $0
        #rm ${dir}gcc/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.

---
 gcc/config/m68k/m68k.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: gcc-4.1/gcc/config/m68k/m68k.h
===================================================================
--- gcc-4.1.orig/gcc/config/m68k/m68k.h
+++ gcc-4.1/gcc/config/m68k/m68k.h
@@ -433,7 +433,7 @@ extern enum reg_class regno_reg_class[];
    assumption that if the class is not ADDR_REGS, then it must be a superset
    of DATA_REGS.  */
 #define LIMIT_RELOAD_CLASS(MODE, CLASS) \
-  (((MODE) == QImode && (CLASS) != ADDR_REGS)	\
+  (((MODE) == QImode && (CLASS) == ADDR_REGS)	\
    ? DATA_REGS					\
    : (CLASS))
 
